api/marshaling.go

```diff
@@ -0,0 +1,359 @@
+		messages: configkit.EntityMessageNames{
+			Produced: message.NameRoles{},
+			Consumed: message.NameRoles{},
+		},
+		handlers: configkit.HandlerSet{},
+	}
+
+	var err error
+	out.identity, err = unmarshalIdentity(in.GetIdentity())
+	if err != nil {
+		return nil, err
+	}
+
+	out.typeName = in.GetTypeName()
+	if out.typeName == "" {
+		return nil, errors.New("application type name is empty")
+	}
+
+	var indices []nameRole
+	for _, nrIn := range in.GetMessages() {
+		nrOut, err := unmarshalNameRole(nrIn)
+		if err != nil {
+			return nil, err
+		}
+		indices = append(indices, nrOut)
+	}
+
+	for _, hIn := range in.GetHandlers() {
+		hOut, err := unmarshalHandler(indices, hIn)
+		if err != nil {
+			return nil, err
+		}
+
+		out.handlers.Add(hOut)
+
+		for n, r := range hOut.MessageNames().Produced {
+			out.messages.Produced[n] = r
+		}
+
+		for n, r := range hOut.MessageNames().Consumed {
+			out.messages.Consumed[n] = r
+		}
+	}
+
+	return out, nil
+}
+
+// marshalHandler marshals a handler config to its protobuf representation.
+//
+// It populates app.Messages with NameRole pairs for each of the handler's
+// consumed/produced messages.
+func marshalHandler(
+	app *pb.Application,
+	indices map[message.Name]uint32,
+	in configkit.Handler,
+) (*pb.Handler, error) {
+	out := &pb.Handler{}
+
+	var err error
+	out.Identity, err = marshalIdentity(in.Identity())
+	if err != nil {
+		return nil, err
+	}
+
+	out.TypeName = in.TypeName()
+	if out.TypeName == "" {
+		return nil, errors.New("handler type name is empty")
+	}
+
+	out.Type, err = marshalHandlerType(in.HandlerType())
+	if err != nil {
+		return nil, err
+	}
+
+	names := in.MessageNames()
+	out.Produced, err = marshalNameRoles(app, indices, names.Produced)
+	if err != nil {
+		return nil, err
+	}
+
+	out.Consumed, err = marshalNameRoles(app, indices, names.Consumed)
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// unmarshalHandler unmarshals a handler configuration from its protocol buffers
+// representation.
+func unmarshalHandler(
+	indices []nameRole,
+	in *pb.Handler,
+) (configkit.Handler, error) {
+	out := &handler{
+		messages: configkit.EntityMessageNames{
+			Produced: message.NameRoles{},
+			Consumed: message.NameRoles{},
+		},
+	}
+
+	var err error
+	out.identity, err = unmarshalIdentity(in.GetIdentity())
+	if err != nil {
+		return nil, err
+	}
+
+	out.typeName = in.GetTypeName()
+	if out.typeName == "" {
+		return nil, errors.New("handler type name is empty")
+	}
+
+	out.handlerType, err = unmarshalHandlerType(in.GetType())
+	if err != nil {
+		return nil, err
+	}
+
+	out.messages.Produced, err = unmarshalNameRoles(indices, in.GetProduced())
+	if err != nil {
+		return nil, err
+	}
+
+	out.messages.Consumed, err = unmarshalNameRoles(indices, in.GetConsumed())
+	if err != nil {
+		return nil, err
+	}
+
+	return out, nil
+}
+
+// marshalNameRoles marshals a message.NameRoles collection for a handler into
+// protocol buffers application representation.
```

Comment:
This is the shared generic part, so it probably shouldn't mention "application"?
```suggestion
// protocol buffers representation.
```

Also (minor), you've got mixed wording of _"'protobuf'/'protocol buffers' representation"_ in your mashal/unmarshal doc block. This doc block should match those.

Reply:
Nah this ones a bit weird, it marshals the messages used by a **handler**, but it puts information into the protobuf representation of the **application**.  It's just the way the protobuf data is organised to avoid repeating the names of messages that are used by multiple handlers.